### ISSUE TITLE
Elide saving callee-saved registers if function doesn't return

### ISSUE THREAD

**Issue by atn34** [llvm-mos/llvm-mos#190]
Consider this c program
```
char g(char, char);
  
void foo() {
        char x0 = 0;
        char x1 = 0;
        char x2 = 0;
        for (;;) {
                x0 = g(x1, x2);
                x1 = g(x0, x1);
                x2 = g(x1, x0);
        }
}
```
for which llvm-mos generates (with -O3)
```
        .text
        .file   "test.c"
        .globl  foo                             ; -- Begin function foo
        .type   foo,@function
foo:                                    ; @foo
; %bb.0:
        lda     mos8(__rc20)
        pha
        lda     mos8(__rc21)
        pha
        ldx     #0
        stx     mos8(__rc21)
        ldx     #0
.LBB0_1:                                ; =>This Inner Loop Header: Depth=1
        lda     mos8(__rc21)
        jsr     g
        sta     mos8(__rc20)
        ldx     mos8(__rc21)
        jsr     g
        ldx     mos8(__rc20)
        sta     mos8(__rc21)
        jsr     g
        tax
        jmp     .LBB0_1
.Lfunc_end0:
        .size   foo, .Lfunc_end0-foo
                                        ; -- End function
        .ident  "clang version 15.0.0 (https://www.github.com/llvm-mos/llvm-mos.git d8b5ba2066d1ee73305fbedab1a209163202b355)"
        .section        ".note.GNU-stack","",@progbits
        .addrsig
```

It seems to me that since this function doesn't return and we never restore the callee-saved registers __rc20 and __rc21, that we don't need to save them in the first place.

For some additional context I'm playing with [dodo](http://www.dodolabs.io/) where your whole game basically must fit in 0x1f00 bytes, so minimizing code size is my primary interest. Also I'm hoping to get around to adding dodo as a supported platform for llvm-mos-sdk if that's welcome!

**Comment by mysterymath** [comment]
Sounds reasonable, since we don't need stack unwinders to work on the 6502. Shouldn't be too difficult to fix.

**Comment by mysterymath** [comment]
It looks like there may be support already in LLVM for this: see determineCalleeSaves in TargetFrameLoweringImpl.cpp. It may just be a matter of turning it on.

**Comment by gandhi56** [comment]
`determineCalleeSaves()` is already being called in the mos-zero-page-alloc pass. The allocations seem to be related to multiple calls to `g`.

**Comment by gandhi56** [comment]
IR (1):
```
define void @test_callee_saved_regs() local_unnamed_addr #0 {
entry:
  br label %for.cond
for.cond:                                         ; preds = %for.cond, %entry
  %x1.0 = phi i8 [ 0, %entry ], [ %call, %for.cond ]
  %call = tail call zeroext i8 @g(i8 noundef zeroext %x1.0) #2
  %call2 = tail call zeroext i8 @g(i8 noundef zeroext %x1.0) #2 ; <== line 7
  br label %for.cond
}

declare dso_local zeroext i8 @g(i8 noundef zeroext) local_unnamed_addr #1
```

ISA (1) after running `./bin/llc -O3 test.ll -o -`:
```
test_callee_saved_regs:                 ; @test_callee_saved_regs
; %bb.0:                                ; %entry
	lda	mos8(__rc20)
	pha
	lda	mos8(__rc21)
	pha
	ldx	#0
.LBB0_1:                                ; %for.cond
                                        ; =>This Inner Loop Header: Depth=1
	stx	mos8(__rc21)
	txa
	jsr	g
	sta	mos8(__rc20)
	lda	mos8(__rc21)
	jsr	g
	ldx	mos8(__rc20)
	jmp	.LBB0_1
.Lfunc_end0:
```
; %bb.0:                                ; %entry
	lda	#0
.LBB0_1:                                ; %for.cond
                                        ; =>This Inner Loop Header: Depth=1
	jsr	g
	jmp	.LBB0_1
.Lfunc_end0:
Erasing line 7 generates the following code:
```

```

**Comment by mysterymath** [comment]
Don't think that's quite right; the reason that its not saving the callee-saved-registers rc20 and rc21 if there's only one function call is that there's no need to use those registers at all. The only reason the compiler uses them is to keep the arguments to the second call alive across the first call.

The observation here is that the function here is stated never to return, there's no reason to actually *save* rc20 and rc21. They can be used freely, since any value that the caller may have placed in them can intrinsically never again be read.

The reason I referred to determineCalleeSaves() is that the default implementation checks for NoReturn functions: https://github.com/llvm-mos/llvm-mos/blob/main/llvm/lib/CodeGen/TargetFrameLoweringImpl.cpp#L118

But the optimization there checks "enableCalleeSaveSkip", https://github.com/llvm-mos/llvm-mos/blob/main/llvm/lib/CodeGen/TargetFrameLoweringImpl.cpp#L33, which just returns false by default. To turn this on for llvm-mos, we'd want to make it return true for our target, and run the result through the example and the LLVM and end-to-end test suites, and see what happens.

**Comment by gandhi56** [comment]
Opened a PR at #227

**Comment by mysterymath** [comment]
Closed by PR #227.
